peerstorage.go

```diff
@@ -0,0 +1,171 @@
+package lnd
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightningnetwork/lnd/chainntnfs"
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/prometheus/common/log"
+	"sync"
+)
+
+const (
+	maxWipeWindow = 2016
+)
+
+var (
+	peerStorage = []byte("peer-storage")
+)
+
+type blockViewer interface {
+	BestHeight() int32
+}
+
+type kvdbPeerStorage struct {
+	*chainntnfs.BlockEpochEvent
+	db             kvdb.Backend
+	awaitingDelete map[int32]*btcec.PublicKey
+	mapMtx         sync.RWMutex
+	quit           chan interface{}
+	blockViewer    chainntnfs.BestBlockView
+	wg             sync.WaitGroup
+}
+
+type peerStorageConfig struct {
+	chainNotifier chainntnfs.ChainNotifier
+	db            kvdb.Backend
+	blockViewer   chainntnfs.BestBlockView
+}
+
+func newKvdbPeerStorage(c *peerStorageConfig) (*kvdbPeerStorage, error) {
+	epochEvent, err := c.chainNotifier.RegisterBlockEpochNtfn(nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &kvdbPeerStorage{
+		db:              c.db,
+		BlockEpochEvent: epochEvent,
+		blockViewer:     c.blockViewer,
+	}, nil
+}
+
+func (k *kvdbPeerStorage) start() {
+	errChan := make(chan error)
+
+	k.wg.Add(1)
+	go func() {
+		defer k.wg.Done()
+		err := k.garbageCollector()
+		select {
+		case errChan <- err:
+		case <-k.quit:
+		}
+
+	}()
+
+	k.wg.Add(1)
+	go func() {
+		defer k.wg.Done()
+		select {
+		case err := <-errChan:
+			if err != nil {
+				log.Errorf("kvdbPeerStorage experienced "+
+					"error during operation %v", err)
+			}
+		case <-k.quit:
+		}
+	}()
+}
+
+func (k *kvdbPeerStorage) stop() {
+	close(k.quit)
+	k.wg.Wait()
+
+	return
+}
+
+func (k *kvdbPeerStorage) StorePeerData(data []byte,
+	peerPub *btcec.PublicKey) error {
+	// Check if to delete
+
+	return kvdb.Update(k.db, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(peerStorage)
+		if err != nil {
+			return err
+		}
+
+		peerPubBytes := peerPub.SerializeCompressed()
+
+		return bucket.Put(peerPubBytes, data)
+	}, func() {})
+}
+
+func (k *kvdbPeerStorage) RetrievePeerData(peerPub *btcec.PublicKey) ([]byte,
+	error) {
+
+	var data []byte
+	err := kvdb.View(k.db, func(tx kvdb.RTx) error {
+		bucket := tx.ReadBucket(peerStorage)
+
+		peerPubBytes := peerPub.SerializeCompressed()
+
+		data = bucket.Get(peerPubBytes)
+		return nil
+	}, func() {})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (k *kvdbPeerStorage) MarkForDelete(peerPub *btcec.PublicKey) error {
+	k.mapMtx.Lock()
+	defer k.mapMtx.Unlock()
+
+	height, err := k.blockViewer.BestHeight()
+	if err != nil {
+		return err
+	}
+
+	k.awaitingDelete[int32(height+maxWipeWindow)] = peerPub
```

Comment:
There are some edge cases here where if we close our last channel with a peer and then open one before the expiry window happens, we will still delete the peer storage anyway. Realistically we don't even need this call now that I think about it. Every GC loop (every block) we can just query for all of the peers we are currently storing, filter for no active channels, and further filter for expired, and then prune those. We simply need to add that query to the config and then let the GC do its own pruning.

Your implementation is clever, though. I like the idea you had here. I just think it'd be cleaner if we were lazier about cleaning up the peer storage.

Reply:
Okay thanks how about this,
https://github.com/lightningnetwork/lnd/pull/8490#issuecomment-1971579842

Reply:
We can defer our strategy for garbage collection for now. We can afford the space leak. It's 65k per peer and as long as we require them to have a channel with us to store it in the first place we can limit the impact. For now maybe just have a delete method on the PeerStorage that deletes it instantly. This way the PeerStorage won't have any goroutines (this is a good thing).